kimchi/src/circuits/gate.rs

```diff
@@ -342,25 +385,62 @@ impl GateType {
             chacha_final_where.insert((ChaChaFinal, *r));
         }
 
-        vec![
-            (chacha_pattern, chacha_where),
-            (chacha_final_pattern, chacha_final_where),
-        ]
+        let lookups = [
+            (chacha_pattern, chacha_where, Some(GateLookupTable::Xor)),
+            (
+                chacha_final_pattern,
+                chacha_final_where,
+                Some(GateLookupTable::Xor),
+            ),
+        ];
+
+        // Convert from an array of tuples to a tuple of vectors
+        {
+            let mut patterns = Vec::with_capacity(lookups.len());
+            let mut locations_with_tables = Vec::with_capacity(lookups.len());
+            for (pattern, locations, table) in lookups {
+                patterns.push(pattern);
+                locations_with_tables.push(GatesLookupSpec {
+                    gate_positions: locations,
+                    gate_lookup_table: table,
+                });
+            }
+            (patterns, locations_with_tables)
```

Comment:
There should be a type for patterns, it’s too chaotic imo to have all these tuples. 

```rust
Struct Pattern {
  /// what lookups to perform in a row
  Lookups: Vec<JointLookup>
  /// the gate that has these lookups
  Locations: HashSet<(GateType,  …)>
  /// the table used in the lookups
  Table
}
```

Reply:
I’m also wondering if this is the best way to specify the layout of all these lookups. This might be unrelated to your PR, I still find it extremely hard to follow these values and how they are loosely related

Reply:
Imo we should build some sort of abstractions that gate implementors can then use to construct fixed tables, and use table lookups as part of their gates. If the abstraction is clean, it’ll be easier to build on top of that in the future. Wdyt?

Reply:
> What do we really want to get here?
> 
> * Kimchi can have gates that can perform lookups, oddly gates actually don’t encode these, they are encoded outside of the gate itself.

This is kinda inverted; some gates 'require' lookups in order to behave correctly, but the gates themselves do not perform them. I agree that it would be better if this lived alongside the gates.

> * Then a gate can perform lookups on two rows at a time, which is opaque from the outside (and might lead to UB if the next row does something weird)

Again, the gate isn't doing it. There's an assertion in this file that ensures that we don't trigger 2 lookup configurations on the same line.

> * A gate can also perform several lookups, but they all have to be for the same table. Oddly we don’t always carry this information

They don't have to be from the same table. That they currently are simplified things for this PR, but nothing about the lookup's setup cares about the particular table.

> * A lookup is either a single lookup (you want to check for the existence of a special value) or for a joint lookup (your table has more than one column). Oddly, a jointlookup is built on top of a single lookup

We never use single lookup in isolation, and once we add `table_id` properly, we probably never will. We should rename them to something that reflects that 'single' is a linear combination of witness fields to form one value in the lookup, and that 'joint' is the actual entry in the table, comprising one or more 'single' lookups and the table ID thereof.

> * A single lookup, and thus a joint lookup, can be formed from a linear combination of witness columns (not used currently)

I take it you haven't looked at `chacha_final_pattern`...

> Imo we should build some sort of abstractions that gate implementors can then use to construct fixed tables, and use table lookups as part of their gates. If the abstraction is clean, it’ll be easier to build on top of that in the future. Wdyt? I’m scared that if we continue to build on top of the current abstractions the logic will become harder and harder to work on.

I'm going to try and improve things as I work through implementing 'extended lookup tables', but my priority is getting that done rather than iterating on the existing one.